process/block/shardblock.go

```diff
@@ -246,6 +251,95 @@ func (sp *shardProcessor) ProcessBlock(
 	return nil
 }
 
+// checkMetaHeadersValidity - checks if listed metaheaders are valid as construction
+func (sp *shardProcessor) checkMetaHeadersValidityAndFinality(header *block.Header) error {
+	metablockCache := sp.dataPool.MetaBlocks()
+	if metablockCache == nil {
+		return process.ErrNilMetaBlockPool
+	}
+
+	tmpNotedHdr, err := sp.getLastNotarizedHdr(sharding.MetachainShardId)
+	if err != nil {
+		return err
+	}
+
+	currAddedMetaHdrs := make([]*block.MetaBlock, 0)
+	for _, metaHash := range header.MetaBlockHashes {
+		value, ok := metablockCache.Peek(metaHash)
+		if !ok {
+			return process.ErrNilMetaBlockHeader
+		}
+
+		metaHdr, ok := value.(*block.MetaBlock)
+		if !ok {
+			return process.ErrWrongTypeAssertion
+		}
+
+		currAddedMetaHdrs = append(currAddedMetaHdrs, metaHdr)
+	}
+
+	if len(currAddedMetaHdrs) == 0 {
+		return nil
+	}
+
+	sort.Slice(currAddedMetaHdrs, func(i, j int) bool {
+		return currAddedMetaHdrs[i].Nonce < currAddedMetaHdrs[j].Nonce
+	})
+
+	for _, metaHdr := range currAddedMetaHdrs {
+		err := sp.isHdrConstructionValid(metaHdr, tmpNotedHdr)
+		if err != nil {
+			return err
+		}
+		tmpNotedHdr = metaHdr
+	}
+
+	err = sp.checkMetaHdrFinality(tmpNotedHdr, header.Round)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// check if shard headers are final by checking if newer headers were constructed upon them
+func (sp *shardProcessor) checkMetaHdrFinality(header data.HeaderHandler, round uint32) error {
+	if header == nil {
+		return process.ErrNilBlockHeader
+	}
+
+	sortedMetaHdrs, err := sp.getOrderedMetaBlocks(round)
+	if err != nil {
+		return err
+	}
+
+	lastVerifiedHdr := header
+	// verify if there are "K" block after current to make this one final
+	nextBlocksVerified := 0
+	for _, tmpHdr := range sortedMetaHdrs {
+		if nextBlocksVerified >= sp.metaBlockFinality {
+			break
+		}
+
+		// found a header with the next nonce
+		if tmpHdr.hdr.GetNonce() == header.GetNonce()+1 {
```

Comment:
change header.GetNonce to lastVerifiedHdr.GetNonce

Reply:
Done. Good catch!